File: SLAM_zero_to_hero/1_16/ros2/src/slam_tutorial/src/subscriber.cpp
/**
 * ROS2 C++ Subscriber Example
 *
 * Subscribes to robot pose messages and tracks distance traveled.
 */

#include <cmath>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"

class PoseSubscriber : public rclcpp::Node {
public:
    PoseSubscriber()
        : Node("pose_subscriber_cpp"),
          last_x_(0.0), last_y_(0.0),
          total_distance_(0.0),
          first_msg_(true),
          msg_count_(0) {

        // Create subscriber
        subscription_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "/robot/pose", 10,
            std::bind(&PoseSubscriber::callback, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "C++ Pose subscriber started");
    }

private:
    void callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
        double x = msg->pose.position.x;
        double y = msg->pose.position.y;

        // Calculate distance traveled
        if (!first_msg_) {
            double dx = x - last_x_;
            double dy = y - last_y_;
            double distance = std::sqrt(dx * dx + dy * dy);
            total_distance_ += distance;
        }
        first_msg_ = false;

        last_x_ = x;
        last_y_ = y;
        msg_count_++;

        // Log periodically
        if (msg_count_ % 50 == 0) {
            RCLCPP_INFO(this->get_logger(),
                "Received pose: (%.2f, %.2f), total distance: %.2fm",
                x, y, total_distance_);
        }
    }

    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr subscription_;
    double last_x_, last_y_;
    double total_distance_;
    bool first_msg_;
    int msg_count_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<PoseSubscriber>());
    rclcpp::shutdown();
    return 0;
}

File: SLAM_zero_to_hero/1_16/ros2/src/slam_tutorial/slam_tutorial/subscriber.py
#!/usr/bin/env python3
"""
ROS2 Subscriber Example

Subscribes to robot pose messages and tracks distance traveled.
Demonstrates basic ROS2 Python subscriber pattern.
"""

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
import math


class PoseSubscriber(Node):
    def __init__(self):
        super().__init__('pose_subscriber')

        # Create subscriber
        self.subscription = self.create_subscription(
            PoseStamped,
            '/robot/pose',
            self.callback,
            10
        )

        # Track distance
        self.last_x = None
        self.last_y = None
        self.total_distance = 0.0
        self.msg_count = 0

        self.get_logger().info('Pose subscriber started')

    def callback(self, msg):
        x = msg.pose.position.x
        y = msg.pose.position.y

        # Calculate distance traveled
        if self.last_x is not None:
            dx = x - self.last_x
            dy = y - self.last_y
            distance = math.sqrt(dx * dx + dy * dy)
            self.total_distance += distance

        self.last_x = x
        self.last_y = y
        self.msg_count += 1

        # Log periodically
        if self.msg_count % 50 == 0:
            self.get_logger().info(
                f'Received pose: ({x:.2f}, {y:.2f}), '
                f'total distance: {self.total_distance:.2f}m'
            )


def main(args=None):
    rclpy.init(args=args)

    subscriber = PoseSubscriber()

    try:
        rclpy.spin(subscriber)
    except KeyboardInterrupt:
        pass
    finally:
        subscriber.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
